src/coroutine/memoset/prove.rs

```diff
@@ -0,0 +1,354 @@
+use crate::{
+    circuit::gadgets::pointer::AllocatedPtr,
+    error::ProofError,
+    field::LurkField,
+    lem::{pointers::Ptr, store::Store},
+    proof::{
+        nova::{CurveCycleEquipped, E1},
+        supernova::{Proof, PublicParams, SuperNovaPublicParams, C2, SS1, SS2},
+        FrameLike, Prover, RecursiveSNARKTrait,
+    },
+};
+
+use bellpepper_core::{num::AllocatedNum, ConstraintSystem, SynthesisError};
+use once_cell::sync::OnceCell;
+use tracing::info;
+
+use super::query::Query;
+use super::{
+    CoroutineCircuit, LogMemo, MemoSet, Scope, DEFAULT_RC_FOR_QUERY,
+    DEFAULT_TRANSCRIBE_INTERNAL_INSERTIONS,
+};
+
+use nova::{
+    supernova::{
+        error::SuperNovaError, snark::CompressedSNARK, NonUniformCircuit, RecursiveSNARK,
+        StepCircuit,
+    },
+    traits::{
+        snark::{BatchedRelaxedR1CSSNARKTrait, RelaxedR1CSSNARKTrait},
+        Dual as DualEng,
+    },
+};
+use std::marker::PhantomData;
+
+const COROUTINE_ARITY: usize = 12;
+
+type Coroutine<'a, F, Q> = CoroutineCircuit<'a, F, LogMemo<F>, Q>;
+
+#[derive(Debug)]
+pub(crate) struct MemosetProver<'a, F, Q> {
+    /// The number of small-step reductions performed in each recursive step of
+    /// the primary Lurk circuit.
+    reduction_count: usize,
+    transcribe_internal: bool,
+    _phantom: PhantomData<&'a (F, Q)>,
+}
+
+impl<'a, F, Q> MemosetProver<'a, F, Q> {
+    pub(crate) fn new(reduction_count: usize, transcribe_internal: bool) -> Self {
+        Self {
+            reduction_count,
+            transcribe_internal,
+            _phantom: PhantomData,
+        }
+    }
+    pub(crate) fn default() -> Self {
+        Self::new(DEFAULT_RC_FOR_QUERY, DEFAULT_TRANSCRIBE_INTERNAL_INSERTIONS)
+    }
+}
+
+impl<'a, F, Q> NonUniformCircuit<E1<F>> for Coroutine<'a, F, Q>
```

Comment:
Seeing this makes me want to ask: how are you planning to integrate coroutines with LEM? I thought we were going to wrap them with `Frame` like we do for coprocessors, then batch them with `MultiFrame`, which already implements `NonUniformCircuit`

Reply:
Yes, we're gonna integrate it with LEM afterwards. We will have to write a query trait for LEM funcs, I already have it sketched in my head